Enable `axis` support for `tf.unique`

yongtang opened the pull request:
The `axis` support for `Unique` has been Added in PR #12952 (defined in `UniqueV2` ops). The support for `axis` in python version of the `tf.unique` was not enabled yet, due to the API workflow porcess (3 weeks). This fix adds the support for `axis` with `tf.unique` by adds `Unique` to `hidden.txt`, and adds a python wrapper of `tf.unique` to pointing to `UniqueV2`.

This fix addresses part of the issue #15644.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

drpngx commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -1252,6 +1252,52 @@ def sparse_mask(a, mask_indices, name=None):
     return ops.IndexedSlices(out_values, out_indices, a.dense_shape)
 
 
+def unique(x, out_idx=None, name=None, axis=None):
```
Can we do without this wrapper?

yongtang replied:
@drpngx It should be possible if we add a line `    .Input("axis: int64")` to array_ops.cc:
```
REGISTER_OP("Unique")
    .Input("x: T")
    .Output("y: T")
    .Output("idx: out_idx")
    .Attr("T: type")
    .Attr("out_idx: {int32, int64} = DT_INT32")
    .SetShapeFn([](InferenceContext* c) {
      c->set_output(0, c->Vector(InferenceContext::kUnknownDim));
      c->set_output(1, c->input(0));
      return Status::OK();
    })
```

Though that means one additional input for the existing API. Will that be OK?

drpngx replied:
Right, that's the old op. So we need the wrapper just to rename into v2, but why the rest of the code? I don't think it's needed. Then we can just have `unique = gen_array_ops.unique_v2`.

yongtang commented:
Thanks @drpngx. The PR tries to handle two cases. One is `axis = nil` and another is axis is not nil (`0, 1, 2, etc`).

However, the `Input` field does not have a way to handle `not exist` case. (Ref https://stackoverflow.com/questions/42754965/marking-input-as-optional-in-tensorflow).

So this PR tries to use `axis = []` to represent `axis = nil` and use `axis =[0]` (one-element array) to represent `axis = 0`.

I think alternatively there are some other ways to simplify:
1. We could use `Attr` for `axis` then it might be possible to have a `not exist` case.
2. We could create two Ops, one for `UniqueOp` and another for `UniqueWithAxisOp` so that axis = None could be routed to `UniqueOp` and axis = 0 could be routed to  `UniqueWithAxisOp`.

Not sure which way is the best. Maybe we could use `Attr` instead.

yongtang commented:
@drpngx I updated the PR. Now `unique` will be routed to `gen_array_ops._unique` if axis = None and `gen_array_ops._unique_v2` otherwise. `UniqueV2` will take a scalar axis now. That should make the `def unique()` simpler. Please take a look.

drpngx commented:
So `axis=None` is not `axis = []`? Basically an empty tensor.

yongtang commented:
@drpngx In case `None` is passed an ValueError as TensorFlow will try to convert to a Tensor:
```
ValueError: Tried to convert 'axis' to a tensor and failed. Error: None values not supported.
```

drpngx commented:
I see. Can the wrapper do that conversion?

yongtang commented:
@drpngx An easy way is to pass `[]` if axis is `None`, then `[]` will be converted to a 1-D tensor. An explicit conversion is also possible I think.

drpngx commented:
OK, sounds good.

yongtang commented:
@drpngx The PR has been updated. Please take a look.

drpngx commented:
Jenkins, test this please.

yongtang commented:
Thanks @drpngx. The previous Jenkins error was caused by `//tensorflow/tools/api/tests:api_compatibility_test`. This is expected as we switch `tf.unique` from Unique to UniqueV2.

I have run API golden and updated the PR:
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
           --update_goldens True
```

I think the Jenkins should pass now.

drpngx commented:
@vrv does the hiding trigger a compatibility issue?

drpngx commented:
Jenkins, test this please.

vrv commented:
I doubt it, since the python API contract is the same, but you can check with @annarev maybe?

yongtang commented:
@drpngx @vrv The PR switched `tf.unique` in python from Unique to UniqueV2 so there is one extra `axis` for `tf.unique`.

It also hides `Unique` and `UniqueV2`.

Previously `Unique` was not hidden and was used by `tf.unique` implicitly (`gen_array_ops.unique`). 

The `UniqueV2` probably should have been hidden at the first place though it will not have an impact on user I think.

annarev commented:
Yep, it shouldn't cause a compatibility issue.

asimshankar commented:
Thanks for the PR. A few things from API review:

- The `UniqueV2` operation takes an `int64` axis. It should take an `int32` consistent with other operations that take an axis argument. It should be fine to change `UniqueV2` in place since there are no users of it. (And then we'll have to wait another 3 weeks for the Python change)

- Documentation of the `UniqueV2` op in C++ seems incorrect (for example, starts with "1-D tensor " :)

- Update the documentation for `tf.unique` to include examples with the `axis` argument.

yongtang committed:
Change axis from int64 to int32 for UniqueV2.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update python wrapper for `tf.unique`

and hide Unique and UniqueV2

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update test cases as Unique and UniqueV2 are hidden

redirect to gen_array_ops._unique and gen_array_ops._unique_v2

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@asimshankar Thanks. The PR has been updated and v2 switching has been removed. Now this PR does:
1. Change axis from int64 to int32.
2. Update docstring in array_ops.cc
3. Hide both Unique and UniqueV2
4. Create a thin python wrapper to point `tf.unique` to `gen_array_ops._unique`. (Will switch to v2 3 weeks after this PR is merged).

Please take a look and let me know if there are any issues.

yongtang committed:
Update UniqueOp to allow both int32 and int64 axis.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Actually, I think it makes sense to support both int32 and int64 axis. The PR has been updated for that. Please take a look.

martinwicke commented:
(good for API)

But could you add an example to the documentation that shows what happens when axis is set?

yongtang committed:
Update doc string

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update API compatibility/golden with
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
           --update_goldens True
```

and
```
tensorflow/core/api_def/update_api_def.sh
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke. The PR has been pushed with docstring update. Please take a look.

martinwicke committed:
Fix typo

martinwicke commented:
Argh -- I think while this happened, @annarev merged the change which moves the docs to APIDef, creating a nontrivial conflict.

yongtang commented:
@martinwicke Let me take a look and fix the conflict.

annarev commented:
@yongtang and @martinwicke to fix it, you can just remove newly added/updated .Doc() calls.
api_def_*.pbtxt are now the location where docs are stored.

yongtang committed:
Remove `.Doc` from UniqueV2 to resolve merge conflict.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke @annarev. The PR has been rebased and pushed. Please take a look.

drpngx commented:
Jenkins, test this please.

yongtang commented:
@martinwicke @drpngx All tests passed except `cla/google` which seems to stuck. Maybe it requires an override?

martinwicke commented:
Yes, since I added a commit, it got stuck.
